### ISSUE TITLE
Having trouble enabling buildkit for building images on CapRover

### ISSUE THREAD

**Issue by piriye** [caprover/caprover#1582]
Hey guys, I am trying to deploy an app that requires buildkit. I have followed all the instructions here [BuildKit | Docker Documentation](https://docs.docker.com/build/buildkit/), but the app still doesn’t build with it enabled. Any idea what i’m doing wrong here?

<img width="1186" alt="Screenshot 2022-12-06 at 00 09 58" src="https://user-images.githubusercontent.com/7614038/205766698-146799a2-926d-4080-912c-99729c263c28.png">

**Comment by githubsaturn** [comment]
It looks like Docker API doesn't support buildkit 
https://github.com/docker-java/docker-java/issues/1904#issuecomment-1193041563
https://github.com/docker/for-linux/issues/1136
https://github.com/moby/moby/issues/41711

**Comment by deadcoder0904** [comment]
i got this error.

```bash
Step 8/44 : RUN --mount=type=cache,id=pnpm,target=~/.pnpm-store pnpm fetch

{"message":"the --mount option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled"}
the --mount option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled
Build has failed!
----------------------
Deploy failed!
Error: {"message":"the --mount option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled"}
the --mount option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled
```
guess I'll have to remove `--mount` option to deploy on caprover.

**Comment by Cryszon** [comment]
Apparently BuildKit is now available in Docker API according to https://github.com/moby/moby/issues/41711#issuecomment-2160179739.

**Comment by VictorioBerra** [comment]
@githubsaturn Can we get support for this since it is now in the API?

**Comment by githubsaturn** [comment]
Is builtkit backward compatible? If so, why is Docker still defaulting to version 1 (non buildkit)?

**Comment by githubsaturn** [comment]
Anyways, the change itself is very simple. All we need to do is to pass `version: "2"` (as string) here:

https://github.com/caprover/caprover/blob/ca196e51be2df80836ff027a99bb92dde83c4f7f/src/docker/DockerApi.ts#L280-L283

But I want to understand the implications.

**Comment by Cryszon** [comment]
According to [Docker documentation](https://docs.docker.com/build/buildkit/#getting-started) and [Release Notes](https://docs.docker.com/engine/release-notes/23.0/#2300) BuildKit is the default builder since Docker 23.0.0, which was released on 2023-02-01. I'm not sure about incompatibilities, but at least for me personally BuildKit is mandatory due to features it enables.

**Comment by githubsaturn** [comment]
[Docker API defaults to version 1 
](https://docs.docker.com/reference/api/engine/version/v1.47/#tag/Image/operation/ImageBuild)
![image](https://github.com/user-attachments/assets/c7d7c17c-7acf-4e4f-8f15-e36c6ed7c567)

**Comment by githubsaturn** [comment]
> I am not an expert on this stuff but @githubsaturn version gets passed to the modem and the path to the API is built like so https://github.com/apocas/docker-modem/blob/master/lib/modem.js#L144 but only if specified.

Yes, that's why I mentioned above, that `version: "2"` needs to be added to the code I linked above.

**Comment by githubsaturn** [comment]
Please comment here if you're still having problems after upgrading to edge version.

**Comment by clue** [comment]
> This is implemented. I need someone with an actual usecase to [update to edge](https://caprover.com/docs/troubleshooting.html#how-to-use-the-edge-version) and test this 😄

@githubsaturn Thank you very much, can confirm Dockerfiles requiring BuiltKit work just fine after updating to edge, no noticeable regressions on my side.
